File: aqp_client/src/lib/api/workloads.ts
/**
 * Workload Studio API client — drives `/manage/*` through the in-AQP
 * proxy installed by `aqp/api/proxy.py` (or directly to the
 * `aqp_control_plane` micro-project when run in client mode).
 *
 * Mirrors the shape of `aqp_platform_core.runtime.WorkloadRuntime`
 * so the SPA can stay provider-agnostic — docker_compose / kubernetes
 * / aws / azure / gcp / cloudflare all return the same envelope.
 */
import { apiFetch } from "./client";

export interface DeploymentStatus {
  service_id: string;
  provider: string;
  phase: string;
  replicas_desired: number;
  replicas_ready: number;
  image?: string | null;
  namespace?: string | null;
  last_transition_at?: string | null;
  conditions?: Array<Record<string, unknown>>;
  endpoints?: Record<string, string>;
  raw?: Record<string, unknown>;
}

export interface WorkloadExecResult {
  service_id: string;
  namespace?: string | null;
  container?: string | null;
  command: string[];
  stdout: string;
  stderr: string;
  returncode?: number | null;
  elapsed_ms?: number | null;
  started_at?: string | null;
  finished_at?: string | null;
}

export interface WorkloadLogEvent {
  service_id: string;
  namespace?: string | null;
  container?: string | null;
  line: string;
  timestamp?: string | null;
  source: string;
}

export interface ResponseEnvelope<T> {
  status: string;
  data: T;
  error?: { code: string; message: string; details?: Record<string, unknown> } | null;
}

export async function listDeployments(namespace?: string): Promise<DeploymentStatus[]> {
  const qs = namespace ? `?namespace=${encodeURIComponent(namespace)}` : "";
  const env = await apiFetch<ResponseEnvelope<DeploymentStatus[]>>(`/manage/deployments${qs}`);
  return env.data ?? [];
}

export async function getDeployment(
  serviceId: string,
  namespace?: string,
): Promise<DeploymentStatus | null> {
  const qs = namespace ? `?namespace=${encodeURIComponent(namespace)}` : "";
  const env = await apiFetch<ResponseEnvelope<DeploymentStatus>>(
    `/manage/deployments/${encodeURIComponent(serviceId)}${qs}`,
  );
  return env.data ?? null;
}

export async function startDeployment(
  serviceId: string,
  spec: Record<string, unknown>,
): Promise<DeploymentStatus | null> {
  const env = await apiFetch<ResponseEnvelope<DeploymentStatus>>(
    `/manage/deployments/${encodeURIComponent(serviceId)}/start`,
    { method: "POST", body: JSON.stringify({ ...spec, service_id: serviceId }) },
  );
  return env.data ?? null;
}

export async function stopDeployment(
  serviceId: string,
  namespace?: string,
): Promise<DeploymentStatus | null> {
  const qs = namespace ? `?namespace=${encodeURIComponent(namespace)}` : "";
  const env = await apiFetch<ResponseEnvelope<DeploymentStatus>>(
    `/manage/deployments/${encodeURIComponent(serviceId)}/stop${qs}`,
    { method: "POST" },
  );
  return env.data ?? null;
}

export async function scaleDeployment(
  serviceId: string,
  replicas: number,
  namespace?: string,
): Promise<DeploymentStatus | null> {
  const params = new URLSearchParams({ replicas: String(replicas) });
  if (namespace) params.set("namespace", namespace);
  const env = await apiFetch<ResponseEnvelope<DeploymentStatus>>(
    `/manage/deployments/${encodeURIComponent(serviceId)}/scale?${params}`,
    { method: "PATCH" },
  );
  return env.data ?? null;
}

export async function restartDeployment(
  serviceId: string,
  namespace?: string,
): Promise<DeploymentStatus | null> {
  const qs = namespace ? `?namespace=${encodeURIComponent(namespace)}` : "";
  const env = await apiFetch<ResponseEnvelope<DeploymentStatus>>(
    `/manage/deployments/${encodeURIComponent(serviceId)}/restart${qs}`,
    { method: "POST" },
  );
  return env.data ?? null;
}

/**
 * Halt every in-flight workload_runs row on the active control plane.
 * Wired into the topbar KillSwitch (Phase B of the Management Engine
 * plan).
 */
export async function haltAllWorkloads(): Promise<{ halted: number }> {
  return apiFetch<{ halted: number }>("/workloads/halt", { method: "POST" });
}

export async function execInDeployment(
  serviceId: string,
  payload: {
    command: string[];
    container?: string;
    timeout_seconds?: number;
    stdin_b64?: string;
    namespace?: string;
  },
): Promise<WorkloadExecResult | null> {
  const env = await apiFetch<ResponseEnvelope<WorkloadExecResult>>(
    `/manage/deployments/${encodeURIComponent(serviceId)}/exec`,
    { method: "POST", body: JSON.stringify(payload) },
  );
  return env.data ?? null;
}

File: aqp_platform_core/src/aqp_platform_core/models/workloads.py
"""Workload-level wire-format models — actions, run rows, exec, logs.

These types power the extended :class:`InfrastructureProvider` surface
(``exec`` / ``tail_logs`` / ``restart`` / ``rotate_secret``) introduced
by AGENTS rule 45's Management Engine phase. They stay stable across
release cycles since the SPA + Theia desktop + ``rpi_k8s_sdk.aqp``
client all serialise them on the wire.

``WorkloadAction`` + ``WorkloadRun`` + ``WorkloadRunStatus`` are the
shared audit ledger schema used by both the in-process
:class:`aqp_platform_core.runtime.workload.WorkloadRuntime` and the
sidecar :class:`aqp_cp` service. ``aqp_cp.models.audit`` re-exports
them for backwards compatibility.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class WorkloadAction(str, Enum):
    """Set of audited control-plane actions (Management Engine).

    Phase 1 of the control-plane maturation adds ``BUILD_IMAGE``
    (Kaniko in-cluster image builds), ``PROVISION_TENANT`` (per-tenant
    namespace bootstrap), ``HALT`` (kill-switch fan-out), and the
    Terraform IaC actions relocated under the modified rule 42.
    """

    START = "start"
    STOP = "stop"
    SCALE = "scale"
    RESTART = "restart"
    APPLY_CONFIG = "apply_config"
    EXEC = "exec"
    LOGS = "logs"
    DELETE = "delete"
    ROTATE_SECRET = "rotate_secret"
    HALT = "halt"
    # Phase 1 — tenant + build orchestration on the control plane.
    PROVISION_TENANT = "provision_tenant"
    DEPROVISION_TENANT = "deprovision_tenant"
    BUILD_IMAGE = "build_image"
    # Rule-42 relocation — Terraform IaC actions executed by the CP.
    TERRAFORM_PLAN = "terraform_plan"
    TERRAFORM_APPLY = "terraform_apply"
    TERRAFORM_DESTROY = "terraform_destroy"
    TERRAFORM_REFRESH = "terraform_refresh"
    TERRAFORM_IMPORT = "terraform_import"
    TERRAFORM_STATE_PULL = "terraform_state_pull"
    TERRAFORM_VALIDATE = "terraform_validate"
    TERRAFORM_UNLOCK = "terraform_unlock"


class WorkloadRunStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    REJECTED = "rejected"
    HALTED = "halted"


class WorkloadRun(BaseModel):
    """Append-only audit row for a single workload operation.

    ``user_id`` is the JWT ``sub``. ``request_id`` is propagated from
    the inbound ``X-Request-Id`` header when present. ``provider`` is
    the active :class:`InfrastructureProvider` alias. ``target`` is the
    resource identifier (service_id, namespace, etc.).

    AGENTS rule 34: every run-producing flow carries ``experiment_id``
    + ``test_id``; the WorkloadRuntime stamps them when set on the
    ``RequestContext``.
    """

    model_config = ConfigDict(extra="forbid")

    run_id: str = Field(
        default_factory=lambda: str(uuid.uuid4()),
        description="Globally unique id (UUID4).",
    )
    started_at: datetime
    finished_at: datetime | None = None
    status: WorkloadRunStatus = WorkloadRunStatus.PENDING
    action: WorkloadAction
    provider: str
    target: str
    namespace: str | None = None
    user_id: str
    request_id: str | None = None
    org_id: str | None = None
    workspace_id: str | None = None
    experiment_id: str | None = None
    test_id: str | None = None
    payload: dict[str, Any] = Field(
        default_factory=dict,
        description="Sanitised request body — secrets MUST be redacted.",
    )
    result: dict[str, Any] = Field(
        default_factory=dict,
        description="Provider response metadata (not raw secrets).",
    )
    error: str | None = None
    halt_reason: str | None = None
    duration_ms: float | None = None


class WorkloadExecResult(BaseModel):
    """Result of an in-pod / in-container command execution.

    Mirrors :class:`aqp.kubernetes.protocol.PodExecResult` but lives in
    ``aqp_platform_core`` so the micro-project and the monolith share
    one Pydantic schema on the wire. ``returncode`` may be ``None`` on
    backends that don't surface an explicit exit code (Docker SDK
    streaming path); callers should inspect ``stderr`` for failure.
    """

    model_config = ConfigDict(extra="forbid")

    service_id: str
    namespace: str | None = None
    container: str | None = None
    command: list[str]
    stdout: str = ""
    stderr: str = ""
    returncode: int | None = None
    elapsed_ms: float | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None


class WorkloadLogEvent(BaseModel):
    """Single log frame emitted by :meth:`InfrastructureProvider.tail_logs`.

    The Management Engine adapts these to the canonical
    ``{task_id, stage, message, timestamp, **extras}`` payload shape
    required by AGENTS rule 4 before pushing to the WebSocket bus.
    """

    model_config = ConfigDict(extra="forbid")

    service_id: str
    namespace: str | None = None
    container: str | None = None
    line: str
    timestamp: datetime | None = None
    source: str = Field(
        default="stdout",
        description="'stdout' | 'stderr' | provider-specific stream name.",
    )


class SecretRotationResult(BaseModel):
    """Outcome of a :meth:`InfrastructureProvider.rotate_secret` call.

    Never contains the secret value itself — only metadata about the
    rotation (rotation id, backend, version pointer). The Management
    Engine subagent rule forbids logging secret material, and this
    model is the only sanctioned response shape.
    """

    model_config = ConfigDict(extra="forbid")

    service_id: str
    secret_name: str
    backend: str = Field(
        description=(
            "Secret backend identifier — 'k8s_secret', 'aws_secretsmanager', "
            "'azure_keyvault', 'gcp_secretmanager', 'auth0_management_api', "
            "'cloudflare_zero_trust'."
        )
    )
    rotation_id: str | None = Field(
        default=None,
        description="Provider-issued rotation id when available.",
    )
    new_version: str | None = Field(
        default=None,
        description=(
            "Opaque pointer to the new secret version. Operators / agents "
            "can use this to verify a rotation completed, but MUST NEVER "
            "log it alongside the secret value."
        ),
    )
    rotated_at: datetime | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


__all__ = [
    "SecretRotationResult",
    "WorkloadAction",
    "WorkloadExecResult",
    "WorkloadLogEvent",
    "WorkloadRun",
    "WorkloadRunStatus",
]
